File: packages/twilio_flutter_conversations/ios/Classes/Bridge/ConversationsEventEmitter.swift
import Foundation

/// Sends canonical conversation events from iOS to Dart.
final class ConversationsEventEmitter {
  private let flutterApi: TwilioConversationsFlutterApi

  init(binaryMessenger: FlutterBinaryMessenger) {
    flutterApi = TwilioConversationsFlutterApi(binaryMessenger: binaryMessenger)
  }

  func emit(event: ConversationsNativeEvent) {
    Task { @MainActor in
      try? await flutterApi.onNativeEvent(event: event)
    }
  }
}

File: packages/twilio_flutter_conversations/android/src/main/kotlin/com/twilioflutter/conversations/bridge/ConversationsEventEmitter.kt
package com.twilioflutter.conversations.bridge

import com.twilioflutter.conversations.pigeon.ConversationsNativeEvent
import com.twilioflutter.conversations.pigeon.TwilioConversationsFlutterApi
import io.flutter.plugin.common.BinaryMessenger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/** Sends canonical conversation events from Android to Dart. */
class ConversationsEventEmitter(binaryMessenger: BinaryMessenger) {
  private val flutterApi = TwilioConversationsFlutterApi(binaryMessenger)
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

  /** Emits [event] to the Flutter side. */
  fun emit(event: ConversationsNativeEvent) {
    scope.launch {
      flutterApi.onNativeEvent(event)
    }
  }
}
